osu.Game/Overlays/Profile/Header/PreviousUsernamesContainer.cs

```diff
@@ -0,0 +1,181 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Shapes;
+using osu.Framework.Graphics.Sprites;
+using osu.Framework.Input.Events;
+using osu.Game.Graphics;
+using osu.Game.Users;
+using osuTK;
+using System;
+using System.Linq;
+
+namespace osu.Game.Overlays.Profile.Header
+{
+    public class PreviousUsernamesContainer : Container
+    {
+        private const int duration = 200;
+        private const int margin = 10;
+        private const int width = 350;
+        private const int move_offset = 20;
+
+        public readonly Bindable<User> User = new Bindable<User>();
+
+        private readonly ContentContainer contentContainer;
+
+        public PreviousUsernamesContainer()
+        {
+            HoverIconContainer hoverIcon;
+
+            AutoSizeAxes = Axes.Y;
+            Width = width;
+            Children = new Drawable[]
+            {
+                contentContainer = new ContentContainer(),
+                hoverIcon = new HoverIconContainer(),
+            };
+
+            hoverIcon.ActivateHover += () =>
+            {
+                contentContainer.Show();
+                this.MoveToY(-move_offset, duration, Easing.OutQuint);
+            };
+
+            User.BindValueChanged(onUserChanged);
+            Hide();
+        }
+
+        private void onUserChanged(ValueChangedEvent<User> user)
+        {
+            contentContainer.Clear();
+
+            var usernames = user.NewValue.PreviousUsernames;
+
+            if (usernames.Any())
+            {
+                var amount = usernames.Length;
+
+                for (int i = 0; i < amount; i++)
+                {
+                    string text = (i + 1 == amount) ? usernames[i] : $@"{usernames[i]},";
+
+                    contentContainer.Usernames.AddText(new SpriteText
+                    {
+                        Font = OsuFont.GetFont(weight: FontWeight.Bold, italics: true),
+                        Text = text,
+                    });
+                }
+
+                Show();
+                return;
+            }
+
+            Hide();
+        }
+
+        protected override void OnHoverLost(HoverLostEvent e)
+        {
+            base.OnHoverLost(e);
+            contentContainer.Hide();
+            this.MoveToY(0, duration, Easing.OutQuint);
+        }
+
+        private class ContentContainer : VisibilityContainer
+        {
+            private const int header_height = 25;
+            private const int content_padding = 60;
```

Comment:
This is too big, should be 2 * padding (i.e. 30)

Reply:
It's the padding from the left of the whole container, not the icon.
Using 30 will looks like this
![Screenshot_4](https://user-images.githubusercontent.com/22874522/62344271-00ce8600-b4f6-11e9-8b47-c6b1487cd04d.png)

Reply:
Nvm, anyway should be adjusted with the new font size

Reply:
That sounds bad. Consider restructuring this entire component so that you have:

```yaml
new Container()
  Children:
    - new Box() // Background
    - new FillFlowContainer()
      Direction: Horizontal
      Children:
        - new HoverIconContainer()
        - new TextContent()
```

Reply:
Yeah, sounds good. I wasn't sure about this for some reason

Reply:
I remember, why. HoverIconContainer should be separate from the whole container. It should not be included inside of any nested containers because in some cases we need to hide ALL the stuff and in some cases we want to hide All the stuff without that icon.

Reply:
But nothing stops you from calling Show()/Hide() on the background + text.

Reply:
Done